src/click/core.py

```diff
@@ -2027,7 +2027,7 @@ def __init__(
         ] = None,
         autocompletion: t.Optional[
             t.Callable[
-                [Context, t.List[str], str], t.List[t.Union[t.Tuple[str, str], str]]
+                [Context, t.List[str], str], t.Iterable[t.Union[t.Tuple[str, str], str]]
```

Comment:
Don't change this from `List` to `Iterable`, I was deliberate about returning `List`.

Reply:
What's the reason for requiring user code to adhere to the more strict `List` type with their return value instead of `Iterable`?

For the purpose of type checking this is (seemingly) not backwards compatible as code that previously passed now starts being flagged. (generator function used as `autocompletion` function). I understand that this is a new major release, so breaking backwards compatibility is an option. But simultaneously the entire purpose of this intermediate shape of `autocompletion` seems to be a migration path to allow people to move away from that towards `shell_complete`. So this confuses me...